File: frontend/types/evolution.ts
export interface EvolutionStrategy {
  id: string
  user_id: string
  course_id: string | null
  strategy_type: string
  before_value: Record<string, unknown>
  after_value: Record<string, unknown>
  description: string | null
  status: string
  risk_level: string
  evidence: string[]
  previous_strategy_id: string | null
  version_no: number
  created_at: string
  updated_at: string
}

export interface EvolutionEvent {
  id: string
  user_id: string
  course_id: string | null
  trigger_type: string
  focus: string
  input_snapshot: Record<string, unknown>
  strategies_generated: number
  status: string
  error_message: string | null
  created_at: string
}

export interface AnalyzeResponse {
  event_id: string
  strategies_count: number
  strategies: EvolutionStrategy[]
}

File: backend/app/schemas/evolution.py
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class AnalyzeRequest(BaseModel):
    course_id: UUID
    focus: str


class StrategyRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    course_id: UUID | None = None
    strategy_type: str
    before_value: dict[str, Any] = {}
    after_value: dict[str, Any] = {}
    description: str | None = None
    status: str
    risk_level: str
    evidence: list[Any] = []
    previous_strategy_id: UUID | None = None
    version_no: int = 1
    created_at: datetime
    updated_at: datetime


class StrategyApplyRequest(BaseModel):
    strategy_id: UUID


class EventRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    course_id: UUID | None = None
    trigger_type: str
    focus: str
    input_snapshot: dict[str, Any] = {}
    strategies_generated: int = 0
    status: str
    error_message: str | None = None
    created_at: datetime
